### PR TITLE
support refcocog eval

### PR DESCRIPTION
**Pull request by xiexinch**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: configs/_base_/datasets/refcocog.py [hunk 1]
```diff
@@ -64,11 +43,12 @@
...
+        data_prefix=dict(img_path='train2014/'),
         ann_file='refcocog/instances.json',
         split_file='refcocog/refs(umd).p',
         split='test',
-        pipeline=test_pipeline,
-        backend_args=backend_args))
+        pipeline=test_pipeline))
 
 # TODO: set the metrics
+val_evaluator = dict(type='SimpleIoUMetric', label_key='gt_masks')
```
Comment (hhaAndroid) [inline configs/_base_/datasets/refcocog.py:53]: SimpleIoUMetric -> ReferSegIoUMetric

File: configs/_base_/datasets/refcoco.py [hunk 2]
```diff
@@ -68,7 +47,7 @@
         ann_file='refcoco/instances.json',
         split_file='refcoco/refs(unc).p',
         split='testA',  # or 'testB'
-        pipeline=test_pipeline,
-        backend_args=backend_args))
+        pipeline=test_pipeline))
 
-# TODO: set the metrics
+val_evaluator = dict(type='ReferSegIoUMetric', label_key='gt_masks')
```
Comment (hhaAndroid) [inline configs/_base_/datasets/refcoco.py:52]:
```suggestion
val_evaluator = dict(type='RefSegIoUMetric', label_key='gt_masks')
```

File: mmdet/evaluation/metrics/semseg_metric.py [hunk 3]
```diff
@@ -277,3 +277,60 @@ def print_semantic_table(
                                         ret_metrics_class['Acc']):
             print_result[class_name] = {'IoU': iou, 'Acc': acc}
         print_log(print_result, logger)
+
+
+@METRICS.register_module()
+class ReferSegIoUMetric(BaseMetric):
```
Comment (hhaAndroid) [inline mmdet/evaluation/metrics/semseg_metric.py:283]: 单独开一个文件吧，RefSegMetric

File: mmdet/evaluation/metrics/semseg_metric.py [hunk 4]
```diff
@@ -277,3 +277,60 @@ def print_semantic_table(
...
+
+    def __init__(self,
+                 label_key: str,
```
Comment (hhaAndroid) [inline mmdet/evaluation/metrics/semseg_metric.py:286]: label_key 不需要

Comment (hhaAndroid) [reply mmdet/evaluation/metrics/semseg_metric.py:286]: 应该做成和 seg metric 一样，用户只需要输入评估指标的名字就行

File: mmdet/evaluation/metrics/refseg_metric.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+from typing import Sequence
+
+import torch
+from mmengine.evaluator import BaseMetric
+
+from mmdet.registry import METRICS
+
+
+@METRICS.register_module()
+class RefSegIoUMetric(BaseMetric):
```
Comment (hhaAndroid) [inline mmdet/evaluation/metrics/refseg_metric.py:11]: RefSegMetric  因为可能为了会评估bbox

Comment (hhaAndroid) [reply mmdet/evaluation/metrics/refseg_metric.py:11]: metric 可以用户传入，现在应该是 ['ciou','miou']

File: configs/_base_/datasets/refcoco.py [hunk 6]
```diff
@@ -68,7 +48,8 @@
         ann_file='refcoco/instances.json',
         split_file='refcoco/refs(unc).p',
         split='testA',  # or 'testB'
-        pipeline=test_pipeline,
-        backend_args=backend_args))
+        text_mode='concat',
+        pipeline=test_pipeline))
 
-# TODO: set the metrics
+val_evaluator = dict(type='RefSegIoUMetric', metrics=['cIoU', 'mIoU'])
```
Comment (hhaAndroid) [inline configs/_base_/datasets/refcoco.py:54]:
```suggestion
val_evaluator = dict(type='RefSegMetric', metrics=['cIoU', 'mIoU'])
```

File: mmdet/datasets/refcoco.py [hunk 7]
```diff
@@ -38,10 +38,17 @@ def __init__(self,
                  data_prefix,
                  split_file,
                  split='train',
+                 text_mode='random',
                  **kwargs):
         self.split_file = split_file
         self.split = split
 
+        assert text_mode in ['random', 'concat', 'select_first', 'all']
```
Comment (hhaAndroid) [inline mmdet/datasets/refcoco.py:46]: all 这个名字有点不太好。要不换成 original，并且放list的第一位

File: configs/_base_/datasets/refcoco.py [hunk 8]
```diff
@@ -52,8 +32,8 @@
         ann_file='refcoco/instances.json',
         split_file='refcoco/refs(unc).p',
         split='val',
-        pipeline=test_pipeline,
-        backend_args=backend_args))
+        text_mode='concat',
```
Comment (hhaAndroid) [inline configs/_base_/datasets/refcoco.py:35]: 默认我们应该选择 original 还是 concat？ 你觉得哪个更合理

Comment (xiexinch) [reply configs/_base_/datasets/refcoco.py:35]: 感觉默认用Random的吧

File: mmdet/evaluation/metrics/__init__.py [hunk 9]
```diff
@@ -22,5 +23,5 @@
     'VOCMetric', 'LVISMetric', 'CrowdHumanMetric', 'DumpProposals',
     'CocoOccludedSeparatedMetric', 'DumpDetResults', 'BaseVideoMetric',
     'MOTChallengeMetric', 'CocoVideoMetric', 'ReIDMetrics', 'YouTubeVISMetric',
-    'COCOCaptionMetric', 'SemSegMetric'
+    'COCOCaptionMetric', 'SemSegMetric', 'RefSegIoUMetric'
```
Comment (hhaAndroid) [inline mmdet/evaluation/metrics/__init__.py:26]:
```suggestion
    'COCOCaptionMetric', 'SemSegMetric', 'RefSegMetric'
```

File: mmdet/evaluation/metrics/__init__.py [hunk 10]
```diff
@@ -12,6 +12,7 @@
 from .lvis_metric import LVISMetric
 from .mot_challenge_metric import MOTChallengeMetric
 from .openimages_metric import OpenImagesMetric
+from .refseg_metric import RefSegIoUMetric
```
Comment (hhaAndroid) [inline mmdet/evaluation/metrics/__init__.py:15]:
```suggestion
from .refseg_metric import RefSegMetric
```

**Review by hhaAndroid — APPROVED** [review]
